src/main/java/edu/greenblitz/tobyDetermined/subsystems/swerve/SwerveChassis.java

```diff
@@ -0,0 +1,459 @@
+	}
+	
+	/**
+	 * makes sure no module is requested to move faster than possible by linearly scaling all module velocities to comply with the constraint
+	 *
+	 * @param states original velocity states computed from the kinematics
+	 * @return states that create the same ratio between speeds but scaled down
+	 */
+	private static SwerveModuleState[] desaturateSwerveModuleStates(SwerveModuleState[] states) {
+		double desaturationFactor = 1;
+		for (SwerveModuleState state : states) {
+			desaturationFactor = Math.max(desaturationFactor, state.speedMetersPerSecond / RobotMap.Swerve.MAX_VELOCITY);
+		}
+		SwerveModuleState[] desaturatedStates = new SwerveModuleState[states.length];
+		for (int i = 0; i < states.length; i++) {
+			desaturatedStates[i] = new SwerveModuleState(states[i].speedMetersPerSecond / desaturationFactor, states[i].angle);
+		}
+		return desaturatedStates;
+	}
+	
+	public ChassisSpeeds getChassisSpeeds() {
+		return kinematics.toChassisSpeeds(getModuleState(Module.FRONT_LEFT),
+				getModuleState(Module.FRONT_RIGHT),
+				getModuleState(Module.BACK_LEFT),
+				getModuleState(Module.BACK_RIGHT));
+	}
+	
+	public SwerveModulePosition[] getSwerveModulePositions() {
+		return new SwerveModulePosition[]{
+				frontLeft.getCurrentPosition(),
+				frontRight.getCurrentPosition(),
+				backLeft.getCurrentPosition(),
+				backRight.getCurrentPosition()
+		};
+	}
+	
+	public SwerveDriveKinematics getKinematics() {
+		return this.kinematics;
+	}
+	
+	public PigeonGyro getPigeonGyro() {
+		return pigeonGyro;
+	}
+	
+	/**
+	 * moving a single module by module state
+	 */
+	private void setModuleStateForModule(Module module, SwerveModuleState state) {
+		getModule(module).setModuleState(state);
+	}
+	
+	/**
+	 * rotates chassis by percentOutput [-1, 1]
+	 */
+	public void rotateChassisByPower(double percentOutput) {
+		for (int i = 0; i < Module.values().length; i++) {
+			Translation2d translationFromCenter = RobotMap.Swerve.SwerveLocationsInSwerveKinematicsCoordinates[i];
+			getModule(Module.values()[i]).rotateToAngle(Math.atan2(translationFromCenter.getY(), translationFromCenter.getX()) + Math.PI * 0.5);
+			getModule(Module.values()[i]).setLinPowerOnlyForCalibrations(percentOutput);
+		}
+	}
+	
+	/**
+	 * for calibration purposes
+	 */
+	public void rotateModuleByPower(Module module, double power) {
+		getModule(module).setRotPowerOnlyForCalibrations(power);
+	}
+	
+	public void updatePoseEstimationPhotonVision() {
+		poseEstimator.update(getPigeonAngle(), getSwerveModulePositions());
+		Photonvision.getInstance().getUpdatedPoseEstimator().ifPresent((EstimatedRobotPose pose) -> poseEstimator.addVisionMeasurement(pose.estimatedPose.toPose2d(), pose.timestampSeconds));
+	}
+	
+	public void updatePoseEstimationLimeLight() {
+		if(MultiLimelight.getInstance().getFirstAvailableTarget().isPresent() || odometry.getPoseMeters().getTranslation().getDistance(getRobotPose().getTranslation()) < RobotMap.Odometry.MAX_DISTANCE_TO_FILTER_OUT) {
```

Comment:
tal - the statements inside the if are too long, extract them to a a previous line and name them to make logic clearer and more readable.